dataplane/handlers/interface.go

```diff
@@ -213,61 +213,102 @@ func (ni *Interface) reconcile(config *oc.Interface) {
-				return
 			}
 		}
 	}
-	// TODO: refactor this.
-	for _, addr := range config.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
-		configIP := addr.Ip
-		configPL := addr.PrefixLength
-		var stateIP *string
-		var statePL *uint8
-		stateAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp())
-		if stateAddr != nil {
-			stateIP = stateAddr.Ip
-			statePL = stateAddr.PrefixLength
-		}
-
-		if configIP != nil && configPL != nil && (stateIP == nil || *statePL != *configPL) {
-			log.V(1).Infof("Config IP: %v, Config PL: %v. State IP: %v, State PL: %v", addr.GetIp(), addr.GetPrefixLength(), stateAddr.GetIp(), stateAddr.GetPrefixLength())
-			log.V(2).Infof("setting interface %s ip %s/%d", tapName, *configIP, *configPL)
-			if err := kernel.SetInterfaceIP(tapName, *configIP, int(*configPL)); err != nil {
+
+	type addrPair struct {
+		cfgIP, stateIP *string
+		cfgPL, statePL *uint8
+	}
+
+	// Get all state IPs and their corresponding config IPs (if they exist).
+	var maybeDeletedPairs []*addrPair
+	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
+		pair := &addrPair{
+			stateIP: addr.Ip,
+			statePL: addr.PrefixLength,
+		}
+		if pairAddr := config.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.cfgIP = pairAddr.Ip
+			pair.cfgPL = pairAddr.PrefixLength
+		}
+		maybeDeletedPairs = append(maybeDeletedPairs, pair)
+	}
+	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv6().Address {
+		pair := &addrPair{
+			stateIP: addr.Ip,
+			statePL: addr.PrefixLength,
+		}
+		if pairAddr := config.GetSubinterface(0).GetIpv6().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.cfgIP = pairAddr.Ip
+			pair.cfgPL = pairAddr.PrefixLength
+		}
+		maybeDeletedPairs = append(maybeDeletedPairs, pair)
+	}
+
+	for _, pair := range maybeDeletedPairs {
+		// If an IP exists in state, but not in config, remove the IP.
+		if pair.stateIP != nil && pair.statePL != nil && (pair.cfgIP == nil && pair.cfgPL == nil) {
+			log.V(1).Infof("Delete Config IP: %v, Config PL: %v. State IP: %v, State PL: %v", pair.cfgIP, pair.cfgPL, *pair.stateIP, *pair.statePL)
+			log.V(2).Infof("deleting interface %s ip %s/%d", tapName, *pair.stateIP, *pair.statePL)
+			if err := kernel.DeleteInterfaceIP(tapName, *pair.stateIP, int(*pair.statePL)); err != nil {
 				log.Warningf("Failed to set ip address of port: %v", err)
-				return
 			}
 		}
 	}
+
+	// Get all config IPs and their corresponding state IPs (if they exist).
+	var maybeAddedPairs []*addrPair
+	for _, addr := range config.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
+		pair := &addrPair{
+			cfgIP: addr.Ip,
+			cfgPL: addr.PrefixLength,
+		}
+		if pairAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.stateIP = pairAddr.Ip
+			pair.statePL = pairAddr.PrefixLength
+		}
+		maybeAddedPairs = append(maybeAddedPairs, pair)
+	}
```

Comment:
This looks to be a dup? Should just have one but the name is `allInterfacePairs` instead of `maybeAddedPairs`?

Reply:
As discussed this is not a dup. Would `allInterfacePairs` still be an improvement since you can merge the SetInterfaceIP/DeleteInterfaceIP bodies?

Reply:
done